python/pyspark/ml/functions.py

```diff
@@ -106,6 +152,597 @@ def array_to_vector(col: Column) -> Column:
+    input arguments of the `predict_fn` (returned by the `make_predict_fn`) and the input columns to
+    the UDF (returned by the `predict_batch_udf`) at runtime.  Each input column will be converted
+    as follows:
+    - scalar column -> np.ndarray
+    - tensor column + tensor shape -> np.ndarray
+
+    Note that tensor columns in the Spark DataFrame must be represented as a flattened 1-D array,
+    and multiple scalar columns can be combined into a single tensor column using standard PySpark
+    SQL functions like `array()` or `struct()`.
+
+    Example (tensor column):
+
+    Input DataFrame has a single column with a flattened tensor value, represented as an array of
+    float.
+    ```
+    from pyspark.ml.functions import predict_batch_udf
+
+    def make_predict_fn():
+        # load/init happens once per python worker
+        import tensorflow as tf
+        model = tf.keras.models.load_model('/path/to/mnist_model')
+
+        # predict on batches of tasks/partitions, using cached model
+        def predict(inputs: np.ndarray) -> np.ndarray:
+            # inputs.shape = [batch_size, 784]
+            # outputs.shape = [batch_size, 10], return_type = ArrayType(FloatType())
+            return model.predict(inputs)
+
+        return predict
+
+    mnist = predict_batch_udf(make_predict_fn,
+                              return_type=ArrayType(FloatType()),
+                              batch_size=100,
+                              input_tensor_shapes=[[784]])
+
+    df = spark.read.parquet("/path/to/mnist_data")
+    df.show(5)
+    # +--------------------+
+    # |                data|
+    # +--------------------+
+    # |[0.0, 0.0, 0.0, 0...|
+    # |[0.0, 0.0, 0.0, 0...|
+    # |[0.0, 0.0, 0.0, 0...|
+    # |[0.0, 0.0, 0.0, 0...|
+    # |[0.0, 0.0, 0.0, 0...|
+    # +--------------------+
+
+    df.withColumn("preds", mnist("data")).show(5)
+    # +--------------------+--------------------+
+    # |                data|               preds|
+    # +--------------------+--------------------+
+    # |[0.0, 0.0, 0.0, 0...|[-13.511008, 8.84...|
+    # |[0.0, 0.0, 0.0, 0...|[-5.3957458, -2.2...|
+    # |[0.0, 0.0, 0.0, 0...|[-7.2014456, -8.8...|
+    # |[0.0, 0.0, 0.0, 0...|[-19.466187, -13....|
+    # |[0.0, 0.0, 0.0, 0...|[-5.7757926, -7.8...|
+    # +--------------------+--------------------+
+    ```
+
+    Example (scalar column):
+
+    Input DataFrame has a single scalar column, which will be passed to the `predict` function as
+    a 1-D numpy array.
+    ```
+    import numpy as np
+    import pandas as pd
+    from pyspark.ml.functions import predict_batch_udf
+    from pyspark.sql.types import FloatType
+
+    df = spark.createDataFrame(pd.DataFrame(np.arange(100)))
+    df.show(5)
+    # +---+
+    # |  0|
+    # +---+
+    # |  0|
+    # |  1|
+    # |  2|
+    # |  3|
+    # |  4|
+    # +---+
+
+    def make_predict_fn():
```

Comment:
ditto: `make_times_two_fn` and then `times_two_udf` below (please also update other examples)